File: Game/src/main/java/net/winternetwork/bedwars/game/modules/shop/listener/VillagerListener.java
package net.winternetwork.bedwars.game.modules.shop.listener;

import lombok.Getter;
import net.winternetwork.bedwars.game.modules.Modules;
import net.winternetwork.bedwars.game.modules.shop.ShopModule;
import net.winternetwork.bedwars.game.modules.shop.model.menu.MenuView;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.inventory.InventoryOpenEvent;
import org.bukkit.event.inventory.InventoryType;

public class VillagerListener implements Listener {

    @Getter(lazy = true)
    private final MenuView view = Modules.getModule(ShopModule.class).getView();

    @EventHandler(priority = EventPriority.HIGHEST)
    public void openEvent(InventoryOpenEvent e) {
        if (e.getInventory().getType() != InventoryType.MERCHANT) return;
        e.setCancelled(true);

        getView().showInventory(((Player) e.getPlayer()));
    }
}

File: Game/src/main/kotlin/net/winternetwork/bedwars/game/modules/shop/listener/VillagerListener.kt
package net.winternetwork.bedwars.game.modules.shop.listener

import net.winternetwork.bedwars.api.util.inject
import net.winternetwork.bedwars.game.modules.shop.ShopModule
import org.bukkit.entity.EntityType
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.inventory.InventoryOpenEvent
import org.bukkit.event.inventory.InventoryType
import org.bukkit.event.player.PlayerInteractEntityEvent

class VillagerListener : Listener {

    val module: ShopModule by inject()

    val view = module.view

    @EventHandler(priority = EventPriority.HIGHEST)
    fun openEvent(e: InventoryOpenEvent) {
        if (e.inventory.type != InventoryType.MERCHANT) return

        e.isCancelled = true
        view.showInventory(e.player as Player)

    }

    @EventHandler(priority = EventPriority.HIGHEST)
    fun interactEvent(e: PlayerInteractEntityEvent) {
        if (e.rightClicked.type != EntityType.VILLAGER) return

        e.isCancelled = true
        view.showInventory(e.player as Player)
    }
}